File: fraiseql-java/src/main/java/com/fraiseql/core/GraphQLField.java
package com.fraiseql.core;

import java.lang.annotation.*;

/**
 * Marks a field as a GraphQL field within a GraphQL type.
 * Automatically extracts type information from the field's Java type.
 *
 * Usage:
 * <pre>
 * @GraphQLType
 * public class User {
 *     @GraphQLField
 *     public int id;
 *
 *     @GraphQLField(nullable = true)
 *     public String email;
 *
 *     @GraphQLField(name = "created_at")
 *     public String createdAt;
 * }
 * </pre>
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.FIELD)
public @interface GraphQLField {
    /**
     * Optional custom name for the GraphQL field.
     * If not specified, the Java field name is used.
     */
    String name() default "";

    /**
     * Whether this field can be null.
     * Nullable fields are represented as Optional&lt;T&gt; in Java.
     */
    boolean nullable() default false;

    /**
     * Optional description for the GraphQL field.
     */
    String description() default "";

    /**
     * Optional custom GraphQL type name.
     * If not specified, the type is inferred from the Java field type.
     */
    String type() default "";
}

File: fraiseql-php/src/Attributes/GraphQLField.php
<?php

declare(strict_types=1);

namespace FraiseQL\Attributes;

use Attribute;

/**
 * PHP 8 Attribute for defining GraphQL field definitions.
 *
 * Usage:
 * ```php
 * #[GraphQLType(name: 'User')]
 * class User {
 *     #[GraphQLField(type: 'Int', nullable: false)]
 *     public int $id;
 *
 *     #[GraphQLField(type: 'String', nullable: false)]
 *     public string $name;
 *
 *     #[GraphQLField(type: 'String', nullable: true)]
 *     public ?string $email;
 * }
 * ```
 */
#[Attribute(Attribute::TARGET_PROPERTY)]
final readonly class GraphQLField
{
    /**
     * @param string|null $type Optional explicit GraphQL type. Auto-detected from property type if not specified.
     * @param string|null $description Optional field description for schema documentation.
     * @param bool $nullable Whether the field is nullable in GraphQL.
     * @param string|null $resolver Optional custom resolver method name.
     */
    public function __construct(
        public ?string $type = null,
        public ?string $description = null,
        public bool $nullable = false,
        public ?string $resolver = null,
    ) {
    }
}
